Fix warnings found by compiling with -Wdeprecated-enum-float-conversion

ddkilzer opened the pull request:
#### c5050dfe582460a512d647870889778fcdb4604d
<pre>
Fix warnings found by compiling with -Wdeprecated-enum-float-conversion
<a href="https://bugs.webkit.org/show_bug.cgi?id=267206">https://bugs.webkit.org/show_bug.cgi?id=267206</a>
&lt;<a href="https://rdar.apple.com/120610023">rdar://120610023</a>&gt;

Reviewed by Tim Nguyen and Darin Adler.

The fix for the warning is to replace the TextSizeAdjustmentType enum
with helper methods that use internal float constants within the class.
The remaining changes are to add some debug assertions and to update the
code to use the new helper methods.

* Source/WebCore/rendering/style/TextSizeAdjustment.h:
(WebCore::TextSizeAdjustmentType): Remove.
(WebCore::TextSizeAdjustment::autoAdjust): Add.
(WebCore::TextSizeAdjustment::none): Add.
- Add static helper methods to replace enum values.
(WebCore::TextSizeAdjustment::TextSizeAdjustment):
- Add debug assertion when setting a float value to make sure it is not
  negative.
- Add private, overloaded constructor that takes a bool parameter to
  implement autoAdjust() and none().
(WebCore::TextSizeAdjustment::percentage const):
(WebCore::TextSizeAdjustment::multiplier const):
- Add debug assertions that m_value is not negative when calling
  percentage() and multiplier().
(WebCore::TextSizeAdjustment::isAuto const):
(WebCore::TextSizeAdjustment::isNone const):
(WebCore::TextSizeAdjustment::Auto): Add.
(WebCore::TextSizeAdjustment::None): Add.
- Private constant float values to implement autoAdjust() and none().

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueWebkitTextSizeAdjust):
* Source/WebCore/style/StyleResolveForDocument.cpp:
(WebCore::Style::resolveForDocument):

Canonical link: <a href="https://commits.webkit.org/272819@main">https://commits.webkit.org/272819@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eb3f927be0cf4c910cba158cf12a17c3405f3c48

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33124 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11898 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35034 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35761 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29911 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34095 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14266 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9067 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29318 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33599 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10049 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29585 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8745 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8905 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29559 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37093 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30117 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29943 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35006 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9028 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6968 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32861 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10733 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9615 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4266 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9678 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/style/TextSizeAdjustment.h:
```diff
@@ -28,24 +28,34 @@ namespace WebCore {
 class RenderStyle;
 
-enum TextSizeAdjustmentType { AutoTextSizeAdjustment = -1, NoTextSizeAdjustment = -2 };
+enum class TextSizeAdjustmentType : bool { None, Auto };
```
I prefer that we nest this inside the `TextSizeAdjustment` class as the name `Type`, and then refer to it as `TextSizeAdjustment::Type`.

ddkilzer replied:
Done!

darinadler commented on Source/WebCore/rendering/style/TextSizeAdjustment.h:
```diff
@@ -28,24 +28,34 @@ namespace WebCore {
 
 private:
-    float m_value;
+    std::variant<float, TextSizeAdjustment::Type> m_value;
```
This makes this bigger than 32 bits. That does not seem like a good tradeoff.

darinadler replied:
Instead we can still use special values for auto and none, but have those special values be private `constexpr` class members. And not use `std::variant`.

ddkilzer replied:
As noted in the commit log message, `dump-class-layout` reports that this PR changes the `WebCore::TextSizeAdjustment` class from 4 bytes to 8 bytes.

Due to existing padding of 4 bytes after `WebCore::TextSizeAdjustment` in `WebCore::StyleRareInheritedData`, this change does not increase the size of `StyleRareInheritedData`.

```
$ xcrun python3 ./Tools/Scripts/dump-class-layout -a arm64e -b /var/build -c Release WebCore WebCore::TextSizeAdjustment
  +0 <  8> WebCore::TextSizeAdjustment
  +0 <  8>     std::__1::variant<float, WebCore::TextSizeAdjustment::Type> m_value
  +0 <  1>         std::__1::__sfinae_ctor_base<true, true> std::__1::__sfinae_ctor_base<true, true>
  +0 <  1>         std::__1::__sfinae_assign_base<true, true> std::__1::__sfinae_assign_base<true, true>
  +0 <  8>         std::__1::__variant_detail::__impl<float, WebCore::TextSizeAdjustment::Type> __impl_
  +0 <  8>             std::__1::__variant_detail::__copy_assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable> std::__1::__variant_detail::__copy_assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable>
  +0 <  8>                 std::__1::__variant_detail::__move_assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable> std::__1::__variant_detail::__move_assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable>
  +0 <  8>                     std::__1::__variant_detail::__assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type> > std::__1::__variant_detail::__assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type> >
  +0 <  8>                         std::__1::__variant_detail::__copy_constructor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable> std::__1::__variant_detail::__copy_constructor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable>
  +0 <  8>                             std::__1::__variant_detail::__move_constructor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable> std::__1::__variant_detail::__move_constructor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable>
  +0 <  8>                                 std::__1::__variant_detail::__ctor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type> > std::__1::__variant_detail::__ctor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type> >
  +0 <  8>                                     std::__1::__variant_detail::__dtor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable> std::__1::__variant_detail::__dtor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable>
  +0 <  8>                                         std::__1::__variant_detail::__base<std::__1::__variant_detail::_Trait::_TriviallyAvailable, float, WebCore::TextSizeAdjustment::Type> std::__1::__variant_detail::__base<std::__1::__variant_detail::_Trait::_TriviallyAvailable, float, WebCore::TextSizeAdjustment::Type>
  +0 <  4>                                           std::__1::__variant_detail::__union<std::__1::__variant_detail::_Trait::_TriviallyAvailable, 0, float, WebCore::TextSizeAdjustment::Type> __data
  +4 <  4>                                           __index_t __index
Total byte size: 8
Total pad bytes: 0
```

ddkilzer replied:
Also note that the `TextSizeAdjustment::Type` enum class is now uses a `bool` for storage.

ddkilzer replied:
Oops, you said 32 _bits_, not 32 _bytes_.  I misread that earlier.

ddkilzer replied:
Just making this change so I can move on to other tasks.

darinadler commented on Source/WebCore/rendering/style/TextSizeAdjustment.h:
```diff
@@ -28,18 +28,19 @@ namespace WebCore {
 public:
-    constexpr TextSizeAdjustment() : m_value(AutoTextSizeAdjustment) { }
+    static constexpr float Auto = -1;
+    static constexpr float None = -2;
```
To get some of the benefits of type safety we could use named constructors instead. Here’s an example of one way to do it that keeps things encapsulated.
```
public:
    static constexpr auto() { return TextSizeAdjustment(true); }
    static constexpr none() { return TextSizeAdjustment(false); }
    constexpr TextSizeAdjustment(float value) : m_value(value) { ASSERT_UNDER_CONSTEXPR_CONTEXT(value >= 0); }
private:
    static constexpr float Auto = -1;
    static constexpr float None = -2;
    constexpr TextSizeAdjustment(bool isAuto) : m_value(isAuto ? Auto : None) { }
```

ddkilzer replied:
I don't think we can name the method `auto()` since `auto` is a C++ keyword.

Would using `Auto()` and `None()` violate code style guidelines?

darinadler replied:
Yes it would. Could name it `autoAdjustment` if you like. Please double check though because I am not **sure** `auto` is disallowed.

ddkilzer replied:
> Please double check though because I am not sure auto is disallowed.

It is not allowed:

```
Source/WebCore/rendering/style/TextSizeAdjustment.h:33:41: error: cannot combine with previous 'type-name' declaration specifier
    static constexpr TextSizeAdjustment auto() { return TextSizeAdjustment(true); }
                                        ^
Source/WebCore/rendering/style/TextSizeAdjustment.h:33:46: error: expected member name or ';' after declaration specifiers
    static constexpr TextSizeAdjustment auto() { return TextSizeAdjustment(true); }
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
```

ddkilzer replied:
Posted a PR using `autoAdjust()` and `none()` using your approach, Darin.

Also considered using `adjustAuto()` and `adjustNone()`, but I didn't want to over-think this after all the back-and-forth.

darinadler commented on Source/WebCore/rendering/style/TextSizeAdjustment.h:
```diff
@@ -28,18 +28,19 @@ namespace WebCore {
-    constexpr float percentage() const { return m_value; }
-    constexpr float multiplier() const { return m_value / 100; }
+    constexpr float percentage() const { ASSERT(m_value >= 0); return m_value; }
+    constexpr float multiplier() const { ASSERT(m_value >= 0); return m_value / 100; }
```
I think this needs to be `ASSERT_UNDER_CONTEXPR_CONTEXT`.

ddkilzer replied:
I did a Debug build, but the build succeeded.

Does `ASSERT_IN_CONSTEXPR_CONTEXT()` exist today?  Did it literally just get committed this afternoon?  I can't seem to find it in WTF.  Oh, it's `ASSERT_UNDER_CONSTEXPR_CONTEXT()`.  Sorry--grep commands failed me.

darinadler replied:
I fixed my wrong suggestion.

darinadler commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -820,9 +820,9 @@ inline void BuilderCustom::applyValueWebkitTextSizeAdjust(BuilderState& builderS
     auto& primitiveValue = downcast<CSSPrimitiveValue>(value);
     if (primitiveValue.valueID() == CSSValueAuto)
-        builderState.style().setTextSizeAdjust(TextSizeAdjustment(AutoTextSizeAdjustment));
+        builderState.style().setTextSizeAdjust(TextSizeAdjustment(TextSizeAdjustment::Auto));
```
This would be cool with my suggested version:
```
builderState.style().setTextSizeAdjust(TextSizeAdjustment::auto());
```

Commit:
Fix warnings found by compiling with -Wdeprecated-enum-float-conversion
https://bugs.webkit.org/show_bug.cgi?id=267206
<rdar://120610023>

Reviewed by Tim Nguyen and Darin Adler.

The fix for the warning is to replace the TextSizeAdjustmentType enum
with helper methods that use internal float constants within the class.
The remaining changes are to add some debug assertions and to update the
code to use the new helper methods.

* Source/WebCore/rendering/style/TextSizeAdjustment.h:
(WebCore::TextSizeAdjustmentType): Remove.
(WebCore::TextSizeAdjustment::autoAdjust): Add.
(WebCore::TextSizeAdjustment::none): Add.
- Add static helper methods to replace enum values.
(WebCore::TextSizeAdjustment::TextSizeAdjustment):
- Add debug assertion when setting a float value to make sure it is not
  negative.
- Add private, overloaded constructor that takes a bool parameter to
  implement autoAdjust() and none().
(WebCore::TextSizeAdjustment::percentage const):
(WebCore::TextSizeAdjustment::multiplier const):
- Add debug assertions that m_value is not negative when calling
  percentage() and multiplier().
(WebCore::TextSizeAdjustment::isAuto const):
(WebCore::TextSizeAdjustment::isNone const):
(WebCore::TextSizeAdjustment::Auto): Add.
(WebCore::TextSizeAdjustment::None): Add.
- Private constant float values to implement autoAdjust() and none().

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueWebkitTextSizeAdjust):
* Source/WebCore/style/StyleResolveForDocument.cpp:
(WebCore::Style::resolveForDocument):

Canonical link: https://commits.webkit.org/272819@main
